MQTT sensors handling of publishing NaN values

kbullet opened the pull request:
# What does this implement/fix?

Publishing "None" instead of "NaN" values for sensors in MQTT. This is to align with HASS's MQTT changes as it no longer seem to like them NaN values in sensor states and it's spamming the logs each time nan is published e.g.

> ValueError: Sensor sensor.ld2410_sensor_g8_move_energy has device class 'None', state class 'None' unit '%' and suggested precision 'None' thus indicating it has a numeric value; however, it has the non-finite value: 'nan'
> 2024-11-08 15:43:41.169 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved (None)
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/components/mqtt/entity.py", line 856, in _async_process_discovery_update
>     await discovery_update(payload)
>   File "/usr/src/homeassistant/homeassistant/components/mqtt/entity.py", line 1137, in discovery_update
>     self.async_write_ha_state()
>   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1005, in async_write_ha_state
>     self._async_write_ha_state()
>   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1130, in _async_write_ha_state
>     self.__async_calculate_state()
>   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
>     state = self._stringify_state(available)
>             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1011, in _stringify_state
>     if (state := self.state) is None:
>                  ^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 675, in state
>     raise ValueError(
> ValueError: Sensor sensor.ld2410_sensor_g8_still_energy has device class 'None', state class 'None' unit '%' and suggested precision 'None' thus indicating it has a numeric value; however, it has the non-finite value: 'nan'

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- https://github.com/esphome/esphome-docs/pull/4452

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```
mqtt:
  broker: $mqtt_broker
  username: $mqtt_user
  password: $mqtt_password
  publish_nan_as_none: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kbullet committed:
Fix nan issue for engineering mode

kbullet committed:
MQTT handling of nan values to align with Hass

kbullet committed:
Merge branch 'dev' into dev

kbullet committed:
removed whitespace

kbullet committed:
updated formatting

kbullet commented:
@jesserockz, @clydebarrow, wdyt? Since API component seems to handle nan to none correctly.

jesserockz commented on esphome/components/mqtt/mqtt_sensor.cpp:
```diff
@@ -70,6 +70,8 @@ bool MQTTSensorComponent::send_initial_state() {
 }
 bool MQTTSensorComponent::publish_state(float value) {
   int8_t accuracy = this->sensor_->get_accuracy_decimals();
+  if (std::isnan(value))
+    return this->publish(this->get_state_topic_(), "None");
```
```suggestion
  if (std::isnan(value))
    return this->publish(this->get_state_topic_(), "None");
  int8_t accuracy = this->sensor_->get_accuracy_decimals();
```

jesserockz committed:
Update esphome/components/mqtt/mqtt_sensor.cpp

jesserockz reviewed (changes requested):
I originally put this on the beta milestone, but just realised this could be a breaking change for people not using HA. 

I think this needs a configuration option (something like `publish_nan_as_none`) for either the global mqtt config, or on the mqtt sensor schema that defaults to false to keep existing functionality.

kbullet committed:
Merge branch 'esphome:dev' into dev

kbullet committed:
Merge branch 'dev' into dev

kbullet committed:
Added mqtt config to publish nan as none

kbullet committed:
Fix lint + test

kbullet committed:
updated formating

kbullet commented:
Thanks @jesserockz thanks for your review and guidance. After further consideration and as per your suggestion I've updated the PR and added as a config for MQTT. 

Also updated docs https://github.com/esphome/esphome-docs/pull/4452

Cheers!

jesserockz commented on esphome/components/mqtt/__init__.py:
```diff
@@ -296,6 +297,7 @@ def validate_fingerprint(value):
                     cv.Optional(CONF_QOS, default=0): cv.mqtt_qos,
                 }
             ),
+            cv.Optional(CONF_PUBLISH_NAN_AS_NONE): cv.boolean,
```
Default values should go in the schema
```suggestion
            cv.Optional(CONF_PUBLISH_NAN_AS_NONE, default=False): cv.boolean,
```

jesserockz commented on esphome/components/mqtt/__init__.py:
```diff
@@ -449,6 +451,9 @@ async def to_code(config):
 
+    if CONF_PUBLISH_NAN_AS_NONE in config:
+        cg.add(var.set_publish_nan_as_none(config[CONF_PUBLISH_NAN_AS_NONE]))
+
```
```suggestion
    cg.add(var.set_publish_nan_as_none(config[CONF_PUBLISH_NAN_AS_NONE]))

```

kbullet committed:
Apply suggestions from code review

Updated

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kbullet committed:
Merge branch 'dev' into dev

kbullet committed:
Merge branch 'dev' into dev

kbullet commented:
Hi, @jesserockz thanks for the suggestions is there anything else you need from my side for this PR?

kbx81 committed:
Merge branch 'dev' into dev
